core/shared/src/main/scala/zio/Chunk.scala

```diff
@@ -1117,6 +1160,136 @@ object Chunk {
     override def toArray[A1 >: A](n: Int, dest: Array[A1]): Unit = { val _ = vector.copyToArray(dest, n, length) }
   }
 
+  private[zio] object Buff {
+    final def arrayBacked[A](buffer: Buffer): Option[Chunk[A]] =
+      if (!buffer.hasArray()) None
+      else {
+        val pos = buffer.position()
+        val lim = buffer.limit()
+        val cap = buffer.capacity()
+        val arr = buffer.array().asInstanceOf[Array[A]]
+        val off = buffer.arrayOffset()
+
+        val chunk = if (off == 0 && pos == 0 && cap == arr.length && lim == cap) {
+          // best case, reuse the underlying array
+          fromArray(arr)
```

Comment:
We can't use the backing array - gotta copy it. The general pattern for reading from File/Socket channels with NIO is based on reusing the same buffer over and over. So if we're grabbing the backing array for the buffer, we will observe mutations as the buffer is re-used.

Reply:
I thought about this and you are completely right. There is another related question to this. Should we do this too in `Chunk.fromArray`?

Reply:
At least we should document that the Array has to be effectively immutable.

Reply:
We should document that but not change it INO.

Reply:
@iravid @vasilmkd I would fixed those two methods and tests for this because it will be very confusing and finding bugs in bigger application will be nightmare.

Reply:
I'm definitely fixing this here. Not sure about `fromArray`, it's not my call to make.

Reply:
I think I would err on the side of making an extra copy here and trading it for performance.

Reply:
Maybe we could add `fromArrayClone` to `fromArray`, but I feel like in most cases, the array passed in will be "write once". It's not like buffers which are often reused.

Reply:
For arrays I would also prefer that we not copy (by default) and document the caveat.